### PR TITLE
Make estargz compression-algorithm-agnostic and support zstd (a.k.a. zstd:chunked)

### PR DESCRIPTION
**Pull request by ktock**
This is the subset of https://github.com/containerd/stargz-snapshotter/pull/281 but re-designed.

Initially, eStargz is based on gzip compression. But, through zstd:chunked work, it turned out that eStargz is not limited to gzip compression and the same chunking & verifying & prefetching method can be applied to other compression algorithms as well (e.g. zstd).

This commit makes `estargz` pkg configurable and agnostic about compression algorithms. For supporting non-gzip compression, the user must implement `estargz.Decompressor` and `estargz.Compressor` interfaces and must plug them to `estargz` tools (e.g. `estargz.Open` and `estargz.NewWriterWithCompression`). `estargz` also provides test suite that is usable for testing these non-gzip eStargz implementations.

This commit comes with `zstdchunked` pkg that support zstd compression for eStargz (a.k.a. zstd:chunked), based on the above extensibility. `zstdchunked` pkg contains `zstdchunked.Decompressor` and `zstdchunked.Compressor` that allows `estargz` pkg to use zstd compression (i.e. zstd:chunked) instead of gzip.

Layer converter and filesystem now support zstd:chunked leveraging `zstdchunked` pkg. `ctr-remote image optimize` and `ctr-remote image convert` support `--zstdchunked` option that omits zstd-based eStargz and filesystem supports zstd-based eStargz layers by default.

In the future, we can support other compression algorithms (e.g. lz4 used by nydus?) as well.

----

cc @giuseppe @AkihiroSuda

### CODE REVIEW

**Comment by ktock** [comment]
@giuseppe WDYT? I know zstd:chunked is still under discussion (https://github.com/containers/storage/pull/775) and that spec may change in the future. But this implementation will help to make that discussion forward?

**Comment by AkihiroSuda** [comment]
@giuseppe Could you take a look? 🙏

**Comment by giuseppe** [comment]
sorry, I've missed the previous ping.

@ktock sure!  And we can try to make it as close as possible to the estargz TOC to simplify using it in stargz-snaphshotter.

The only difference I am aware of is how the chunking is represented.  Does the current metadata for files chunking work for you?

**Comment by ktock** [comment]
@giuseppe Metadata (TOC) implemented in https://github.com/containers/storage/pull/775 looks good to me.
Could you add the following annotation to the result image for enabling chunk verification by stargz snapshotter?

- `containerd.io/snapshot/stargz/toc.digest` = uncompressed digest of manifest (TOC) (`sha256:deadbeef...`)

**Comment by giuseppe** [comment]
> `containerd.io/snapshot/stargz/toc.digest` = uncompressed digest of manifest (TOC) (`sha256:deadbeef...`)

is it fine if it is named `io.containers.zstd-chunked.manifest-checksum` as it is currently?

**Comment by ktock** [comment]
Containerd currently doesn't allow snapshotter to handle labels that are not prefixed by `containerd.io/snapshot/`. Additionally, `io.containers.zstd-chunked.manifest-checksum` seems stores "compressed" digest of TOC JSON but Stargz Snapshotter currently requires "uncompressed" digest of TOC JSON.

I also think we need standard spec for lazy-pulling-related annotations (currently discussed in https://github.com/opencontainers/image-spec/issues/815).

File: nativeconverter/zstdchunked/zstdchunked.go [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+		uncompressedDesc := &desc
+		// We need to uncompress the archive first
+		if !uncompress.IsUncompressedType(desc.MediaType) {
+			var err error
+			uncompressedDesc, err = uncompress.LayerConvertFunc(ctx, cs, desc)
+			if err != nil {
+				return nil, err
+			}
+			if uncompressedDesc == nil {
+				return nil, errors.Errorf("unexpectedly got the same blob aftr compression (%s, %q)", desc.Digest, desc.MediaType)
```
Comment (AkihiroSuda) [inline nativeconverter/zstdchunked/zstdchunked.go:85]: typo

File: nativeconverter/zstdchunked/zstdchunked.go [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+		if err = w.Commit(ctx, n, "", content.WithLabels(labelz)); err != nil && !errdefs.IsAlreadyExists(err) {
+			return nil, err
+		}
+		if err := w.Close(); err != nil {
+			return nil, err
+		}
+		newDesc := desc
+		if uncompress.IsUncompressedType(newDesc.MediaType) {
+			if images.IsDockerType(newDesc.MediaType) {
+				newDesc.MediaType += ".zstd"
```
Comment (AkihiroSuda) [inline nativeconverter/zstdchunked/zstdchunked.go:173]: Do we really want to support zstd for DockerType?

Comment (ktock) [reply nativeconverter/zstdchunked/zstdchunked.go]: No, we don't. Fixed it to always convert to oci type.

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**